doc/tut3.rst

```diff
@@ -0,0 +1,277 @@
+``newLit``, or with ``quote do:`` expressions. This offers the best low
+level control for the syntax tree generation, but the second option is
+much less verbose.  If you choose to create the syntax tree with calls
+to ``newTree`` and ``newLit`` the macro ``marcos.dumpAstGen`` can help
+you with the verbosity. ``quote do:`` allows you to write the code
+that you want to generate literally, backticks are used to insert code
+from ``NimNode`` symbols into the generated expression. This means
+that you can't use backticks within ``quote do:`` for anything else.
+
+
+Building your first macro
+-------------------------
+
+To give a footstart to writing macros we will show now how to
+implement the ``myDebug`` macro mentioned earlier.  The first thing to
+do is to build a simple example of the macro usage, and then just
+print the argument.  This way it is possible to get an idea of a
+correct argument should be look like.
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  import macros
+
+  macro myAssert(arg: untyped): untyped =
+    echo arg.treeRepr
+
+  let a = 1
+  let b = 2
+
+  myAssert(a != b)
+
+.. code-block::
+
+  Infix
+    Ident "!="
+    Ident "a"
+    Ident "b"
+
+
+From the output it is possible to see that the information that the
+argument is an infix operator (node kind is "Infix"), as well as that the two
+operands are at index 1 and 2. With this information the actual
+macro can be written.
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  import macros
+
+  macro myAssert(arg: untyped): untyped =
+    # all node kind identifiers are prefixed with "nnk"
+    arg.expectKind nnkInfix
+    arg.expectLen 3
+    # operator as string literal
+    let op  = newLit(" " & arg[0].repr & " ")
+    let lhs = arg[1]
+    let rhs = arg[2]
+
+    result = quote do:
+      if not `arg`:
+        raise newException(AssertionError,$`lhs` & `op` & $`rhs`)
+
+  let a = 1
+  let b = 2
+
+  myAssert(a != b)
+  myAssert(a == b)
+
+
+This is the code that will be generated. To debug what the macro
+actually generated, the statement ``echo result.repr`` can be used, in
+the last line of the macro. It is also the statement that has been
+used to get this output.
+
+.. code-block:: nim
+  if not (a != b):
+    raise newException(AssertionError, $a & " != " & $b)
+
+With Power Comes Responsibility
+-------------------------------
+
+Macros are very powerful. A good advice is to use them as little as
+possible, but as much as necessary. Macros can change the semantics of
+expressions, making the code incomprehensible for anybody who does not
+know exactly what the macro does with it.  So whenever a macro is not
+necessary and the same logic can be implemented using templates or
+generics, it is probably better not to use a macro.  And when a macro
+is used for something, the macro should better have a well written
+documentation.  And for all the people who write only perfectly self
+explaining code, for macros the implementation is not documentation
```

Comment:
This sentence doesn't sound right to me. I would try with something like:

*For all the people who claim to write only perfectly self-explanatory code: when it comes to macros, the implementation is not enough for documentation.*